sycl/source/detail/scheduler/commands.cpp

```diff
@@ -102,37 +114,46 @@ std::vector<RT::PiEvent> Command::prepareEvents(ContextImplPtr Context) {
       continue;
     }
     ContextImplPtr EventContext = Event->getContextImpl();
-
+    auto Plugin = Event->getPlugin();
     // If contexts don't match - connect them using user event
     if (EventContext != Context && !Context->is_host()) {
 
       EventImplPtr GlueEvent(new detail::event_impl());
       GlueEvent->setContextImpl(Context);
-
       RT::PiEvent &GlueEventHandle = GlueEvent->getHandleRef();
-      PI_CALL(piEventCreate)(Context->getHandleRef(), &GlueEventHandle);
-      PI_CALL(piEventSetCallback)(Event->getHandleRef(), CL_COMPLETE,
-                                  EventCompletionClbk,
-                                  /*data=*/GlueEventHandle);
-      GlueEvents.push_back(std::move(GlueEvent));
-      Result.push_back(GlueEventHandle);
+      Plugin.call<PiApiKind::piEventCreate>(Context->getHandleRef(),
+                                            &GlueEventHandle);
+      EventImplPtr *GlueEventCopy =
+          new EventImplPtr(GlueEvent); // To increase the reference count by 1.
+      Plugin.call<PiApiKind::piEventSetCallback>(
+          Event->getHandleRef(), CL_COMPLETE, EventCompletionClbk,
+          /*void *data=*/(GlueEventCopy));
+      GlueEvents.push_back(GlueEvent);
+      Result.push_back(std::move(GlueEvent));
       continue;
     }
-    Result.push_back(Event->getHandleRef());
+    Result.push_back(Event);
   }
   MDepsEvents.insert(MDepsEvents.end(), GlueEvents.begin(), GlueEvents.end());
   return Result;
 }
 
 void Command::waitForEvents(QueueImplPtr Queue,
-                            std::vector<RT::PiEvent> &RawEvents,
+                            std::vector<EventImplPtr> &EventImpls,
                             RT::PiEvent &Event) {
-  if (!RawEvents.empty()) {
+
+  if (!EventImpls.empty()) {
     if (Queue->is_host()) {
-      PI_CALL(piEventsWait)(RawEvents.size(), &RawEvents[0]);
+      for (auto &EventImpl : EventImpls) {
```

Comment:
You don't need loop here. Please, call waitForEvents with array of events.

Reply:
done.